### PR TITLE
Use a regex to match on a more precise TODO format

### PR DESCRIPTION
**Pull request by rbehjati**
This PR uses a regular expression to check that TODOs are correctly formatted (i.e., specify an issue number). #1007 is a draft of a more comprehensive check (thanks to @daviddrysdale). The main difference between this PR and #1007 is that: 

- #1007 excludes bash scripts from the check
- #1007 checks that numbers specified in the TODOs are actual GitHub issues

### CODE REVIEW

File: runner/src/check_todo.rs [hunk 1]
```diff
@@ -15,16 +15,27 @@
...
     path: String,
+    todo_pattern: Regex,
 }
 
 impl CheckTodo {
     pub fn new(path: String) -> Box<Self> {
-        Box::new(CheckTodo { path })
+        Box::new(CheckTodo {
+            path,
+            todo_pattern: Regex::new("TODO\\(#[0-9][0-9]*\\)").expect("Could not parse regex"),
```
Comment (daviddrysdale) [inline runner/src/check_todo.rs:31]: Drive-by: `\d+`

Also, there might be stuff of interest in #1007&hellip;

Comment (rbehjati) [reply runner/src/check_todo.rs:31]: Wow! Nice. What is your plan for #1007? I could not go to sleep without a regex here, but I feel better now knowing that there is #1007 😄

Comment (daviddrysdale) [reply runner/src/check_todo.rs]: I *might* get back to #1007 now that `runner` is more stable, but it will need a bunch of merging and I'm unlikely to get to it in the short term, so I'd suggest carrying on with your plan.  

(I may also just drop #1007 &ndash; the core idea of checking online issue status is quite brittle because the GitHub API rate limits kick in quickly).

File: runner/src/check_todo.rs [hunk 2]
```diff
@@ -15,16 +15,48 @@
...
     path: String,
+    todo_pattern: Regex,
 }
 
 impl CheckTodo {
     pub fn new(path: String) -> Box<Self> {
-        Box::new(CheckTodo { path })
+        Box::new(CheckTodo {
+            path,
+            todo_pattern: Regex::new(r"TODO\(#\d+\)").expect("Could not parse regex"),
```
Comment (tiziano88) [inline runner/src/check_todo.rs:32]: I think this is re compiling the same regex for each instance of this struct. It seems more efficient to create the regex just once with once_cell or lazy_static instead.

File: runner/src/check_todo.rs [hunk 3]
```diff
@@ -15,16 +15,48 @@
 //
 
 use super::*;
+use regex::Regex;
+use std::cmp::max;
 
 /// A [`Runnable`] command that checks for the existence of todos in the codebase with no associated
```
Comment (tiziano88) [inline runner/src/check_todo.rs:21]: Update the doc to mention that it only looks for todos on commented lines

File: runner/src/check_todo.rs [hunk 4]
```diff
@@ -15,16 +15,48 @@
...
+                comment_depth += line.matches("/*").count() as i32;
+            }
+            // This will include lines annotated with `#` in `.rs` files, which should not be a
+            // problem.
+            if comment_depth > 0 || line.trim().starts_with("//") || line.trim().starts_with('#') {
+                lines.push(line.to_string());
+            }
+            comment_depth = max(
+                0,
+                comment_depth - (line.matches("*/").count() - line.matches("/*/").count()) as i32,
```
Comment (tiziano88) [inline runner/src/check_todo.rs:56]: Why is the match for `/*/` necessary? It should be subsumed by the First one?

File: runner/src/check_todo.rs [hunk 5]
```diff
@@ -15,16 +15,48 @@
...
+        })
+    }
+
+    fn is_invalid_todo_line(&self, line: &str) -> bool {
+        line.split_whitespace()
+            .filter(|word| word.contains("TODO"))
+            .any(|word| !self.todo_pattern.is_match(word))
+    }
+
+    fn get_comment_lines(file_content: String) -> Vec<String> {
```
Comment (rbehjati) [inline runner/src/check_todo.rs:48]: @juliettepretot (I cannot reply to your original comment about multiline comments any more!) 
I've added this function to include multiline comments in the check for `TODO`s. Let me know if there are any important cases that I am missing. Or if you can think of a way to simplify this function :)

File: runner/src/check_todo.rs [hunk 6]
```diff
@@ -15,16 +15,27 @@
...
     path: String,
+    todo_pattern: Regex,
 }
 
 impl CheckTodo {
     pub fn new(path: String) -> Box<Self> {
-        Box::new(CheckTodo { path })
+        Box::new(CheckTodo {
+            path,
+            todo_pattern: Regex::new("TODO\\(#\\d+\\)").expect("Could not parse regex"),
```
[the remark this answers is not in the document]
Comment (juliettepretot) [reply runner/src/check_todo.rs]: Oh that seems nice. I should caution I don't have strong opinions though, was largely just curious. :)

File: runner/src/check_todo.rs (hunk 2 above)
Comment (tiziano88) [reply runner/src/check_todo.rs]: Sorry I didn't mean a literal `once_cell` (which as you figured works but is not very convenient to use for a lazy initialization), rather something like

https://github.com/project-oak/oak/blob/3247b75d335729467ccca375e95f15590c8030e0/runner/src/main.rs#L63

Comment (rbehjati) [reply runner/src/check_todo.rs]: Thanks. Fixed.

File: runner/src/check_todo.rs (hunk 3 above)
Comment (tiziano88) [reply runner/src/check_todo.rs:25]: Just curious, why is the mutex needed here?

Comment (rbehjati) [reply runner/src/check_todo.rs:25]: Too much copy pasting! Sorry! It is not needed!

File: runner/src/check_todo.rs [hunk 7]
```diff
@@ -35,28 +83,41 @@ impl Runnable for CheckTodo {
...
+    }
+
+    #[test]
+    fn comment_lines_are_found_correctly() {
+        let file_content =
+            std::fs::read_to_string("src/check_todo.rs").expect("could not read file");
+        let comment_lines = CheckTodo::get_comment_lines(file_content);
+        println!("{:?}", comment_lines);
+        assert_eq!(comment_lines.len(), 32);
+    }
```
Comment (juliettepretot) [inline runner/src/check_todo.rs:122]: I like this test, it's cute that the code checks the file it exists in.

Comment (juliettepretot) [reply runner/src/check_todo.rs:122]: Might be worth to check a file that has multi-line comments though.

File: runner/src/check_todo.rs (hunk 5 above)
Comment (juliettepretot) [reply runner/src/check_todo.rs:45]: It seem good to me. I kind of do wonder if we really need the `comment_line` check at all though. At this point we're basically implementing our own code parser, that parses files for comments and even supports nested comments. It's nice engineering, but I suspect some of my past colleagues would've called it over-engineered. ;) 

The original code checked for for `"TODO"` regardless of whether it lives it in a comment, and it seemed like that worked fine? 

I personally wouldn't add comment-parsing unless we really need it, but I also don't have a strong opinion on it. :)

Comment (rbehjati) [reply runner/src/check_todo.rs:45]: Yes. I agree. I am not sure about keeping this part either, because of the exact same reasons :) 
I'll change it back to checking for TODOs in every line.
